staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go

```diff
@@ -0,0 +1,32 @@
+// +build windows
+
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+const (
+	// Avoid failed to run Kubelet: bad socketPath,
+	// must be an absolute path: /var/lib/kubelet/device-plugins/kubelet.sock
+	// https://github.com/kubernetes/kubernetes/issues/93262
+
+	// DevicePluginPath is the folder the Device Plugin is expecting sockets to be on
+	// Only privileged pods have access to this path
+	// Note: Placeholder until we find a "standard path"
+	DevicePluginPath = "c:\\var\\lib\\kubelet\\device-plugins\\"
```

Comment:
OS-specific constants are pretty unusual ... it means code that references this constant but is not running on the target OS will get the wrong value

Reply:
I understand but how can I change that. by using `goruntime.GOOS != "windows"` in  https://github.com/kubernetes/kubernetes/blob/5a529aa3a0dd3a050c5302329681e871ef6c162e/pkg/kubelet/cm/devicemanager/manager.go#L134

https://github.com/kubernetes/kubernetes/blob/6aa4c7eca83b74076ae24afde5edff7d932b854e/cmd/kubeadm/app/util/runtime/runtime.go#L62

Reply:
Device plugin paths must be absolute path, see https://github.com/kubernetes/kubernetes/blob/5a529aa3a0dd3a050c5302329681e871ef6c162e/pkg/kubelet/cm/devicemanager/manager.go#L134.

However, the definition of "absolute path" is different on Windows and Linux, see the different implementations of `filepath.IsAbs`, see https://golang.org/src/path/filepath/path_windows.go versus https://golang.org/src/path/filepath/path_unix.go.

Therefore, it is unavoidable that the DevicePluginPath is different under Windows and Linux.

Reply:
it looks like DevicePluginPath is only referenced by KubeletSocket, and KubeletSocket is only referenced by NewManagerImpl in-tree. If we needed a different default path per-platform, something like this would be clearer to me:

```diff
diff --git a/pkg/kubelet/cm/devicemanager/manager.go b/pkg/kubelet/cm/devicemanager/manager.go
index 5d1925f9458..14dc3be1347 100644
--- a/pkg/kubelet/cm/devicemanager/manager.go
+++ b/pkg/kubelet/cm/devicemanager/manager.go
@@ -22,6 +22,7 @@ import (
 	"net"
 	"os"
 	"path/filepath"
+	"runtime"
 	"sort"
 	"sync"
 	"time"
@@ -125,7 +126,11 @@ func (s *sourcesReadyStub) AllReady() bool          { return true }
 
 // NewManagerImpl creates a new manager.
 func NewManagerImpl(numaNodeInfo cputopology.NUMANodeInfo, topologyAffinityStore topologymanager.Store) (*ManagerImpl, error) {
-	return newManagerImpl(pluginapi.KubeletSocket, numaNodeInfo, topologyAffinityStore)
+	socketPath := pluginapi.KubeletSocket
+	if runtime.GOOS == "windows" {
+		socketPath = pluginapi.KubeletSocketWindows
+	}
+	return newManagerImpl(socketPath, numaNodeInfo, topologyAffinityStore)
 }
 
 func newManagerImpl(socketPath string, numaNodeInfo cputopology.NUMANodeInfo, topologyAffinityStore topologymanager.Store) (*ManagerImpl, error) {
```

Reply:
Yes, ideally all paths are the same under Windows and Unix.

However, a) the overall convention of Kubernetes using hard-wired absolute paths, together with b) the meaning of absolute paths being different on Unix and Windows unavoidably leads to the conclusion that we need different path definitions on Unix and Windows.

And in this case I don't see how this is critical anyway, since I don't see any realistic use-case to use `DevicePluginPath` in a cross-compiling situation as you are mentioning.

Reply:
Thanks @liggitt, I changed my PR to add your feedback

Reply:
> OS-specific constants are pretty unusual ... it means code that references this constant but is not running on the target OS will get the wrong value

My understanding here was: only the builds targeting Windows will reference the constants in `constants_windows.go` while Linux builds will reference the corresponding constants in `constants_linux.go` / `constants_unix.go` based on: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants_windows.go and  https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta1/defaults_windows.go  as well as in Go libraries: https://github.com/golang/go/blob/master/src/os/path_windows.go#L8

@liggitt can you explain a bit why/how this does not work. I assume the compiler will fail the build if a targeted platform does not have all the constants defined that the platform agnostic code is referring to?

Reply:
> OS-specific constants are pretty unusual ... it means code that references this constant but is not running on the target OS will get the wrong value

@ddebroy I think @liggitt's concern here is roughly that because we use a single name across platforms for this constant, it may not be obvious when we wind up with code that depends on properties of a specific platform's constant value.

As opposed to having distinct constants for each platform.

If we already have platform specific implementations referencing the constant, why not reference a platform specific constant instead ...?

 > it looks like DevicePluginPath is only referenced by KubeletSocket, and KubeletSocket is only referenced by NewManagerImpl in-tree. If we needed a different default path per-platform, something like this would be clearer to me:

This comment outlines this before showing a specific proposed implementation ...